aten/src/ATen/core/ivalue_inl.h

```diff
@@ -797,16 +799,14 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
   std::vector<std::function<void(void)>> callbacks_;
   std::exception_ptr eptr_;
 
-  // Extra fields needed to optionally support CUDA devices. They will only be
-  // used if impl_ != nullptr.
-
   // An upcast pointer to a virtual class which allows us to manipulate events,
   // streams, ... in a generic way, without an explicit dependency on CUDA.
-  const c10::impl::DeviceGuardImplInterface* const impl_ = nullptr;
+  const c10::impl::VirtualGuardImpl impl_;
```

Comment:
I think `VirtualGuardImpl` might be a more idiomatic way than `DeviceGuardImplInterface`, so I went for it.